src/backend/distributed/sql/udfs/get_missing_time_partition_ranges/latest.sql

```diff
@@ -0,0 +1,264 @@
+    /*
+     * Check distinct partition interval count for the given table.
+     *
+     * If it is 0, that means we are creating the first partition.
+     * We must use the given partition_interval.
+     *
+     * If it is 1, that means all partitions cover the same interval.
+     * We must use that interval and it must be equal to given partition_interval, if it is given.
+     *
+     * If it is more than 1, that means partitions cover different intervals.
+     * We must error out, as partitions with different intervals are not supported
+     *
+     * Note that, to_value and from_value is equal to '', if default partition exist.
+     * To skip them, additional checks are added to the query.
+     *
+     * TODO: Should we lock parent table to not to create partition in parallel?
+     */
+    SELECT
+    COUNT(DISTINCT to_value::timestamptz - from_value::timestamptz)
+    INTO distinct_partition_interval_count
+    FROM time_partitions
+    WHERE parent_table = table_name AND to_value <> '' AND from_value <> '';
+
+    IF distinct_partition_interval_count = 0 THEN
+        table_partition_interval := partition_interval;
+
+        IF partition_interval IS NULL THEN
+            RAISE 'must specify a partition_interval when there are no partitions yet';
+        END IF;
+
+        /*
+         * Decide on the current_range_from_value of the initial partition according to interval of the table.
+         * Since we will create all other partitions by adding intervals, truncating given start time will provide
+         * more intuitive interval ranges, instead of starting from start_from directly.
+         * TODO: Check truncate for quarter
+         */
+        IF table_partition_interval < INTERVAL '1 hour' THEN
+            current_range_from_value = date_trunc('minute', start_from);
+        ELSIF table_partition_interval < INTERVAL '1 day' THEN
+            current_range_from_value = date_trunc('hour', start_from);
+        ELSIF table_partition_interval < INTERVAL '1 week' THEN
+            current_range_from_value = date_trunc('day', start_from);
+        ELSIF table_partition_interval < INTERVAL '1 month' THEN
+            current_range_from_value = date_trunc('week', start_from);
+        ELSIF table_partition_interval = INTERVAL '3 months' THEN
+            current_range_from_value = date_trunc('quarter', start_from);
+        ELSIF table_partition_interval < INTERVAL '1 year' THEN
+            current_range_from_value = date_trunc('month', start_from);
+        ELSE
+            current_range_from_value = date_trunc('year', start_from);
+        END IF;
+
+        current_range_to_value := current_range_from_value + table_partition_interval;
+
+    ELSIF distinct_partition_interval_count = 1 THEN
+        SELECT
+        DISTINCT to_value::timestamptz - from_value::timestamptz
+        INTO table_partition_interval
+        FROM time_partitions
+        WHERE parent_table = table_name AND to_value <> '' AND from_value <> '';
```

Comment:
do we need this extra query? It seems we already determined there is only 1 partition interval, so checking it against the first partition after the query below should do.